### ISSUE TITLE
Capabilities of the underlying loop

### ISSUE THREAD

**Issue by bwoebi** [async-interop/event-loop#16]
The Uv extension exposes functions for watching on signals and to execute operations on files.
Other loops may expose signal handling through pcntl.

a) I assume signal handlers will just throw an exception if you try to register one via onSignal() while unsupported? … Do we want to expose a way to actually access that information on whether it is supported without trying and catching?
b) In general loops are handled by one underlying loop resource (uv, ev) - do we want to be able to expose what type of loop exactly and the underlying loop actually is?

Regarding b), in Amp, it is simple, we have three implementations and can just check for the actual instance and have one function `getLoop()` (not in the interface) on the `UvReactor`. This can be used by our `amphp/file` library to actually operate on it.

But if we are going to have multiple implementations, we need to have a way to access that data (loop type, loop resource if existing).

I'm not really sure how to shape that though? Maybe a function `getLoop()` returning an array `[$extension, $loop]` … so that it would be `["uv", $the_uv_loop]` or `["native", null]`?
I'm really open to any better suggestion here!

**Comment by AndrewCarterUK** [comment]
I don't think we will allow an implementation to not support `onSignal()`. The C `select()` call is woken by signals and alarms so I think it's trivial to support signals even using a stream select loop in PHP. I know Icicle does and I've had to write similar code for PHPFastCGI.

B) really comes back to the discussion in the other thread. Async interop won't be involved in actually implementing the event loop, so the most we can do here is provide an API to retrieve the active driver.

**Comment by bwoebi** [comment]
a) if you don't have pcntl installed and not an Uv-driver you are actually pretty much out of luck ?!

b) I mean, if we provide an interoperable way for event loops, we also need an interoperable way to access its underlyings. Else we'll end up with Amp/file just being compatible with the Amp implementation of the Uv Driver, but nothing else. It'd then have to fall back to native (blocking) behavior, which definitiely isn't what you want if you have a React Uv driver for example.

**Comment by kelunik** [comment]
Is the filesystem the only thing that ever needs these? In that case, we might just add something to the interface. Otherwise, this won't ever work in an interoperable way.

**Comment by AndrewCarterUK** [comment]
I'm assuming pcntl is only not installed on Windows machines, as Windows doesn't have unix signal handling capabilities?

In which case **cough** ~~fuck windows~~ **cough**, erm, we should probably add some mechanism for testing functionality on top of throwing an exception.

A quick suggestions:

``` php
EventLoop::supports(EventLoop::SIGNAL_HANDLING);
```

Then we could also have

``` php
class FeatureNotSupportedException extends \RuntimeException
{
    public function getFeature(); // ...
}
```

**Comment by kelunik** [comment]
Suggestion: `UnsupportedFeatureException`

`pcntl` can also be missing, because you're using a `disable-all` build of PHP, like I usually do.

**Comment by bwoebi** [comment]
UnsupportedFeatureException is fine … in case it is actually tried.
But IMHO we should also have a way which allows checking capabilities directly (e.g. at top-level to immediately abort if e.g. signal watching isn't available) instead of throwing.

**Comment by sagebind** [comment]
Oops, I actually missed this thread...

@AndrewCarterUK That is actually a spot-on idea; I really like it! I'd have `supports()` just return a boolean so that features can be tested ahead of time, like @bwoebi mentioned. Of course, it would be a little strange of an API if there is only one optional feature defined. What else would there be other than signal handling that could be optional?

**Comment by kelunik** [comment]
Polling for exceptional reads, see #19.

**Comment by trowski** [comment]
It makes sense to have a method for checking if signal handling is enabled on the loop, but beyond that I think code requiring a particular feature would have to just use `EventLoop::get()` (or similar) to get the loop instance and see if it implements a particular interface.

**Comment by bwoebi** [comment]
I wouldn't check for a concrete, capabilities may depend on more than just the reactor, but possibly also on the OS or on an additional extension.
Using a method for that in any case makes things simpler.

**Comment by trowski** [comment]
Sure, using a method makes sense, but the only use-case I can see is signal handling. Defining a constant like `EventLoop::FILE_HANDLING` doesn't make a lot of sense if `EventLoop` doesn't include file handling methods.

**Comment by bwoebi** [comment]
As @kelunik mentioned, out of band reads.

Obviously doesn't make sense for file reads, if we don't specify an API here.

**Comment by trowski** [comment]
So you're suggesting if we want to support out-of-band watchers, such as an `onException` watcher, support for it would be indicated through a call such as `EventLoop::supports(EventLoop::OUT_OF_BAND)`?

**Comment by bwoebi** [comment]
basically that.

**Comment by sagebind** [comment]
To implement that, we could add `EventLoopDriver::getSupportedFeatures()` to simplify and centralize feature processing as well. `supports` could just compare the given feature against all the ones the driver supports.

**Comment by WyriHaximus** [comment]
I like the idea of `EventLoopDriver::getSupportedFeatures()` and `supports`. But are these going to be string constants implementers can define themselves or are we going to define them in the specs and release new specs for when there is the need for new one (the latter sounds aweful tbh).

**Comment by sagebind** [comment]
I was thinking integer constants, but they could also be strings. Something like `SIGNAL_HANDLING` and `OUT_OF_BAND` would be defined as part of the spec since it is tied to the driver interface.

**Comment by kelunik** [comment]
All of these constants have to be tied to the spec, otherwise they don't make sense, if every vendor introduces their own constants.

**Comment by WyriHaximus** [comment]
> I was thinking integer constants

With integer constants it would be possible to store all the supported features in one integer and use bitwise operations to store and retrieve supported features.

> All of these constants have to be tied to the spec, otherwise they don't make sense, if every vendor introduces their own constants.

Let me rephrase that: How do we handle new constants after the spec is finalized? Release a new minor version of the spec. The thing I'de hate to see is a new feature popping up and we have no way of adding it to the spec later on.

**Comment by kelunik** [comment]
Why do we need a new major version? A new minor one is fine.

**Comment by WyriHaximus** [comment]
> Why do we need a new major version? A new minor one is fine.

Minor is exactly what I said :wink:

**Comment by kelunik** [comment]
Oops, it's late. Yes, that's the way I'd go.

**Comment by AndrewCarterUK** [comment]
Anyone want to submit a PR for this?

I'm away at a conference this weekend but if not I'll do it when I head back - seems like there is good agreement here.

**Comment by kelunik** [comment]
@AndrewCarterUK What for? For `UnsupportedFeatureException` and the optional feature support API?

I think we still don't have a solution for the main issue: Accessing the underlying event loop resource e.g. for file handling.

**Comment by AndrewCarterUK** [comment]
@kelunik - The `UnsupportedFeatureException()` and the `supports()`

I'm not sure I like `getSupportFeatures()`. If that was the name, I feel you'd need to list everything that was supported (delays, alarms, normal streams, etc...).

I prefer the ability to ask if specific features are there and get a response (`supports()`) - and maybe allow it to accept a variable number of arguments.

The obvious feature is signal handling and I don't see any reason not to add out of bound data too (but that might be best in a later PR as it requries a new method `onOutOfBound()` or similar).

**Comment by AndrewCarterUK** [comment]
I'm not sure we need to be specify the loop driver type (uv, stream select, etc...). What we're creating here is an abstraction layer.

If anyone does want to do driver specific things they can use the `get()` method to get the underlying driver and test with `instanceof` (which they would need to do anyway before manipulating).

**Comment by kelunik** [comment]
> The obvious feature is signal handling and I don't see any reason not to add out of bound data too (but that might be best in a later PR as it requries a new method onOutOfBound() or similar).

We need a solution here anyway, because we can't simply add methods later for other optional features.

> I'm not sure I like getSupportFeatures(). If that was the name, I feel you'd need to list everything that was supported (delays, alarms, normal streams, etc...).

How about `getOptionalFeatures()`? That way only the accessor needs to implement `supports` to figure it out, the driver could just return an array.

> If anyone does want to do driver specific things they can use the get() method to get the underlying driver and test with instanceof (which they would need to do anyway before manipulating).

The `get()` method? I don't see one. But we probably don't need to standardize it, because `instanceof` will be called first anyway.

**Comment by AndrewCarterUK** [comment]
1) Agreed.

2) Sounds good.

3) The `get()` method is here: https://github.com/async-interop/event-loop/pull/17/files#diff-cdc07fcba9000e5df0e709e2ff024127R40

**Comment by kelunik** [comment]
PR #32 is a thing now. I removed `getOptionalFeatures` and went with `supports` instead, otherwise we have to determine support for all features instead of maybe just one.

**Comment by bwoebi** [comment]
PR #32 has the problem that it doesn't solve files.
We really need also a capability of exposing the underlying loop.

I don't like it too much, but e.g. you have two file components which each are coupled to the respective UvLoop implementation. Now, you cannot use the two independent libraries at the same time (in a non-blocking way).
Also, we need to be able to know what type of Loop it is (libuv, libevent, native)?

**Comment by kelunik** [comment]
> We really need also a capability of exposing the underlying loop.

We have `Loop::get`. The driver can then have a method to get the internal loop without it being in the interface. But that makes file system access probably implementation dependent again, right.

Maybe introduce an interface for each driver? `UvLoopDriver`, `NativeDriver`, ...? If a new extension appears, we can update the spec and add another driver type.

**Comment by bwoebi** [comment]
Exactly that's the issue …

The idea of using different interfaces for the different drivers is interesting and probably worth pursuing.
I'd be interested on feedback from the others here.

**Comment by bwoebi** [comment]
But at that point one may wonder again if it should really be interfaces or just directly implementations. There's not much which changes between the loop implementations of a specific type, except that the one implementation may be more performant than the others.
Project specific additions (e.g. projects may want to retain the old way for a while) - can be emulated by extending and calling parent methods.

**Comment by kelunik** [comment]
@bwoebi You're right, but you can still implement a new driver. Just implement no interface at all or the `NativeDriver` (maybe we should just kill that one), other packages such as the ones for filesystem access will just fall back to the native one. Once the implementation is fine, it can be added to the spec.

**Comment by assertchris** [comment]
I feel, given [we should be validating against `LoopDriver` for factories](https://github.com/async-interop/event-loop/pull/30/files#r63295120), that the `LoopDriver` should be the lowest defined interface. Folks can get access to the drivers through `Loop::get()` without us having to define three (?) seperate interfaces for three (?) separate implementations.

**Comment by bwoebi** [comment]
@assertchris the point is that we cannot check against a specific driver … but code using e.g. amphp/file shall be still able to run [in a non-blocking way] with e.g. an Icicle uv driver. And thus we need some interoperability there.

**Comment by assertchris** [comment]
Oh! I understand. It's not that the interop group will suggest multiple Uv drivers but that Icicle/AMPHP/React etc could. It's not that easy to see from https://github.com/async-interop/event-loop/pull/37 since it doesn't describe any Uv-specific methods...

**Comment by bwoebi** [comment]
Yeah, I don't really like the approach with the interfaces … but I don't know of any better approach :-/

**Comment by WyriHaximus** [comment]
It looks weird, was thinking about a sort of enum holding different options, as for example: https://github.com/WyriHaximus/reactphp-child-process-pool/blob/master/src/Options.php But not sure if it would work here

**Comment by assertchris** [comment]
What about an interface for `FileLoopDriver` that defines the Filesystem operations Uv can do without tying it to Uv?

**Comment by WyriHaximus** [comment]
That could work, AFAIK Uv is the only one doing it, but a new one might emerge so it is future proof

**Comment by kelunik** [comment]
@assertchris Files are currently the only thing I think, but there might be other primitives later? Don't know.

> Yeah, I don't really like the approach with the interfaces … but I don't know of any better approach :-/

What's the issue with it?

**Comment by assertchris** [comment]
If we need to create the interfaces to contain certain driver-specific behaviours, why don't we instead create interfaces for those behaviours? 

Edit: then `supports()` could check if `"files" => $driver instanceof FileLoopDriver`

**Comment by kelunik** [comment]
Because those things are tied to the underlying loop, but they don't need to be implemented by the loop ifself. It's perfectly fine to have a loop driver separate from the filesystem handling.

@WyriHaximus Eio is another possible driver, but that's another thing, because that one works with threads and isn't tied to any loop.

**Comment by WyriHaximus** [comment]
@kelunik I'm aware of EIO, spent a significant amount of time working with it, didn't mention it because it's not an event loop :). It has one but it doesn't expose it so we can't tie into it

**Comment by trowski** [comment]
Let's keep the event loop interface simple and not worry about interoperability for signals, file handling, etc. and eliminate the `supports()` method.

To allow interoperability of signal and file handling, we can have separate standards extending the event loop interface that add methods for signal handling, file handling, or anything else that a particular loop implementation may be capable of doing.

``` php
// Just examples, names aren't meant to be suggestions.

class SignalLoop
{
    public static function onSignal($signo, callable $callback, $data = null);
    {
        $loop = Loop::get();

        if ($loop instanceof SignalLoopDriver) {
            throw new UnsupportedFeatureException();
        }

        return $loop->onSignal($signo, $callback, $data);
    }
}

interface SignalLoopDriver extends LoopDriver
{
    public function onSignal($signo, callable $callback, $data = null);
}
```

**Comment by kelunik** [comment]
@trowski  I think that's a bad idea. Keeping it simple is for me just having one interface.

**Comment by trowski** [comment]
Then what does `Loop::supports(Loop::FEATURE_FILE_HANDLING)` mean without file handling methods on `Loop` and `LoopDriver`? Without defining those, checking for file handling support is meaningless. To keep this interface from getting bloated, I would prefer to move support of other features to a different interface.

**Comment by kelunik** [comment]
> `Loop::supports(Loop::FEATURE_FILE_HANDLING)`

It won't exist, because it doesn't make sense.

It's also the reason why I suggested using driver interfaces instead.

**Comment by trowski** [comment]
Hypothetically, suppose another extension using libuv as the underlying loop is released, we'd have to add a `NewUvLoop` interface to allow interoperability between packages implementing `LoopDriver` using this underlying loop extension. If instead an interface such as `FileLoopDriver` was available based on supported features, implementers of `LoopDriver` written using this new underlying loop extension could also implement `FileLoopDriver` and immediately have the same interoperability benefits of any of the other loop implementations.

**Comment by kelunik** [comment]
@trowski The issue is that this makes the driver then not only responsible for its basic loop functionality, but it has to implement the complete file handling instead.

**Comment by kelunik** [comment]
Regardless of that, we can't drop `supports`, as interfaces are compile time, but support might be runtime dependent, such as OS or a missing `pcntl` extension.

**Comment by trowski** [comment]
@kelunik True, an object implementing `LoopDriver` may also implement complete file handling methods to implement a `FileLoopDriver` interface, but only because those methods are tied to the underlying loop extension. File operations will be dependent on the loop, even if the underlying loop was wrapped in a different object. Neither situation is optimal, but if an operation is dependent on the loop, personally I'd rather see it as a method of the loop.

File handling packages will need to provide several drivers (for example, to support EIO if available). One of these drivers could check for a `LoopDriver` implementing `FileLoopDriver` and defer file operations to the event loop.

**Comment by kelunik** [comment]
@trowski The issue is, that we have to standardize the file system access API then, something which isn't required for interoperability.

But I agree, there are issues with the interfaces, because there might be a UV v2.

**Comment by AndrewCarterUK** [comment]
Not sure how related this is to previous discussion:

Regarding driver interfaces, it strikes me as a step in the wrong direction (towards coupling to implementations and away from abstractions).

Unless these interfaces are going to offer different functionality (and even then we now have `supports()`) being able to test them with an `instanceof` doesn't actually offer anything. Just because a driver uses `libuv` doesn't mean it supports the operation you were intending. You still need to test that it's a specific `libuv` driver to perform any off-spec operations.

I'm not opposed, in principle, to different interfaces for different functionality.

**Comment by kelunik** [comment]
@AndrewCarterUK Our goal isn't really abstraction, but rather interoperability.

> libuv doesn't mean it supports the operation you were intending. You still need to test that it's a specific libuv driver to perform any off-spec operations.

If you have the underlying loop resource, you don't need support in the driver. That's how Amp does it currently.

**Comment by assertchris** [comment]
@kelunik Abstraction is integral to interoperability. I don't understand what you're trying to say.

**Comment by kelunik** [comment]
@assertchris Something like those driver interfaces is still interoperable, even if it doesn't abstract anything. But yes, we would have to extend the spec every time there's a new driver type, which is bad.

**Comment by AndrewCarterUK** [comment]
@kelunik - I can see a bit more what you're getting at here. I'm torn because these methods open an incentive for people to create code that isn't interoperable, but by not having them we potentially close the door on functionality.

I'll have a look through again this evening and see if I can source some external opinions.

**Comment by rdlowrey** [comment]
Put me on record as strongly opposed to `supports()` ... Why have an interface contract if you then have to check which parts of the abstraction actually apply? That defeats the whole purpose.

**Comment by AndrewCarterUK** [comment]
@rdlowrey - Understand totally, but the other options appear to be:
- Remove onSignal
- Don't allow the standard to be used in environments that don't support signal handling

**Comment by kelunik** [comment]
@rdlowrey Because signal handling isn't available on all systems. That way you can check for support before it throws and maybe warn the user or do something else.

**Comment by rdlowrey** [comment]
Yeah, I totally get the "why" of `supports()` ... I just disagree with the logic. The early amp implementations had an `interface EventLoop` (or equivalent) with an `interface SignalHandlingEventLoop extends EventLoop` that implementations could implement if they supported the `onSignal()` event.

I subsequently merged the two into a single loop and if a user tried to use `onSignal()` in the one environment where signals were unavailable I preferred to throw a `YourOperatingSystemIsTerribleException` ... I really feel like that's the only sane option. If an application _needs_ signal handling then you need an OS or PHP environment that supports it. There's nothing preventing you from doing this:

``` php
<?php
try {
    EventLoop::onSignal(SIG_FOO, $myCallable);
} catch (CrippledWindowsException $e) {
    // oh snap, we didn't have to create a superfluous API
}
```

IMHO the underlying operating system's inability to handle signals in very specific scenarios _should_ throw; I feel like this is a better approach than being stuck in `if` conditional hell. The real issue I see with `supports()` to expose signal handling availability is this:

> Signal handling capability is a feature of the php execution environment, not the `EventLoop` API.

To me, and this is probably subjective, the existence of `EventLoop::supports()` undermines the very idea of an interface API contract. There's only one very specific case where the environment just doesn't support it and IMHO that doesn't warrant adding unnecessary methods to the interface. The PHP object model already solves this problem via interface extension or the `try/catch` model. 

FWIW you can do signal handling in windows with libuv _or_ libevent ... there are lots of options. It's only in the very basic windows environment without the requisite extensions where you can't. All that said, it wouldn't be the end of the world if `supports()` were a thing. This would just necessitate the inclusion of standard constants that one might pass to it on the loop interface as well.

I think the bigger issue is exposing constant values representing the signal numbers on a per-system basis. If there's no `ext/pcntl` extension available you can't use those. If you're using `pecl/libevent` you just have to pass integers directly. I exposed the relevant constants on a per-operating-system basis in the uv extension but it sucks to not have a unified set of constants to use. Ideally we'd come up with some sane way to represent the signal numbers across event loops with signal handling capability ...

If anyone has suggestions on that it would be helpful.

**Comment by kelunik** [comment]
@rdlowrey: Of course, it _does_ throw if that feature is not available. `supports` just ensures you don't have to create a dummy handler on application startup to check for availability. That way you can immediately error out.

**Comment by bwoebi** [comment]
@rdlowrey In addition to @kelunik there may be signal handlers which just do some non-essential (or just more user-friendly) cleanup e.g. in SIGINT … these are optional and shall just not be set then. That's why we need _some_ way to determine availability (apart from exception catching).

**Comment by bwoebi** [comment]
After talking to @trowski I feel like #48 is absolutely the cleanest and ideal way to go in this problem for getting the underlying resource/object.

**Comment by sagebind** [comment]
@bwoebi Agreed. This is probably the simplest and best solution.

**Comment by sagebind** [comment]
After reviewing this thread and re-thinking the problem, I've changed my stance. I agree with what @rdlowrey said earlier as the only sane and clean option, and I'd like to share my reasoning.

Other than implementing event loops themselves, when have you ever "kind of" needed to handle process signals? A `supports()` method would let you check to see if it is supported ahead of time, like so:

``` php
if (Loop::supports(Loop::SignalsFeature)) {
    Loop::onSignal(SIGUSR1, function () {
        /* ... */
    });
} else {
    // Um, now what?
}
```

What are you going to do if signals aren't supported? To what end are we trying to avoid an exception? If you need to use signals and they aren't supported, there's nothing more you can do to _get_ signal handling. Signal handling not being supported then is clearly an error in your library / app, when clearly you need it.

I'm not trying to say that we shouldn't allow graceful failure, we can still do that with exceptions:

``` php
try {
    Loop::onSignal(SIGUSR1, function () {
        /* ... */
    });
} catch (LoopFeatureNotSupportedException $e) {
    // Fail gracefully here
}
```

As you can see, this has the same if/else structure, without dealing with `supports()`, extra interfaces, or anything complex. If a driver does not support a feature, it should throw an exception.

**Comment by bwoebi** [comment]
I'm not sure on the issue. Pretty much indifferent here.

There's the advantage of supports() to cleanly indicate support (Also for testing at startup `die("Need signal handling!");` without actually setting a signal handler … but well, you can also set and cancel immediately).
The disadvantages are, as you say, that we always need manual error handling here. And an extra function.

In this particular case using try/catch might be fine.

**Comment by kelunik** [comment]
> Other than implementing event loops themselves, when have you ever "kind of" needed to handle process signals?

Uhm, never in event loop themselfs. Always if you want to ensure a clean shutdown, e.g. in Aerys.
